src/meta/src/stream/stream_graph.rs

```diff
@@ -565,156 +596,265 @@ impl BuildActorGraphState {
+        let existing_locations = self.build_locations(external_locations);
 
-        // Serialize the graph
-        let stream_graph = stream_graph
-            .into_iter()
-            .map(|(fragment_id, actors)| {
-                let fragment = self.fragment_graph.seal_fragment(fragment_id, actors);
-                let fragment_id = fragment_id.as_global_id();
-                (fragment_id, fragment)
+        // Extract the new dispatchers from the external changes.
+        let dispatchers = external_changes
+            .iter()
+            .map(|(actor_id, change)| {
+                (
+                    actor_id.as_global_id(),
+                    change.new_downstreams.values().cloned().collect(),
+                )
             })
             .collect();
 
-        Ok(stream_graph)
+        Ok(ActorGraphBuildResult {
+            graph,
+            building_locations,
+            existing_locations,
+            dispatchers,
+        })
     }
 
-    /// Build actor graph from fragment graph using topological order. Setup dispatcher in actor and
-    /// generate actors by their parallelism.
-    fn build_actor_graph(
-        &self,
-        id_gen: GlobalActorIdGen,
-        ctx: &mut CreateStreamingJobContext,
-    ) -> MetaResult<BuildActorGraphState> {
-        let mut state = BuildActorGraphState::new(id_gen);
+    /// Build actor graph for each fragment, using topological order.
+    fn build_actor_graph(&self, id_gen: GlobalActorIdGen) -> MetaResult<ActorGraphBuildStateInner> {
+        let mut state = ActorGraphBuildState::new(id_gen);
 
         // Use topological sort to build the graph from downstream to upstream. (The first fragment
         // popped out from the heap will be the top-most node in plan, or the sink in stream graph.)
         for fragment_id in self.fragment_graph.topo_order()? {
-            // Build the actors corresponding to the fragment
-            self.build_actor_graph_fragment(fragment_id, &mut state, ctx)?;
+            self.build_actor_graph_fragment(fragment_id, &mut state)?;
         }
 
-        Ok(state)
+        Ok(state.finish())
     }
 
+    /// Build actor graph for a specific fragment.
     fn build_actor_graph_fragment(
         &self,
         fragment_id: GlobalFragmentId,
-        state: &mut BuildActorGraphState,
-        ctx: &mut CreateStreamingJobContext,
+        state: &mut ActorGraphBuildState,
     ) -> MetaResult<()> {
-        let current_fragment = self
-            .fragment_graph
-            .get_fragment(fragment_id)
-            .unwrap()
-            .clone();
+        let current_fragment = self.fragment_graph.get_fragment(fragment_id);
+
+        // First, add or record the actors for the current fragment into the state.
+        let (distribution, actor_ids) = match current_fragment {
+            // For building fragments, we need to generate the actor builders.
+            EitherFragment::Building(current_fragment) => {
```

Comment:
It's possible we are visiting the upstream `Materialize` (existing) fragment now, as the topological order is on the complete graph. So we need to distinguish between them.